docs(cn): translate learn/add-react-to-an-existing-project into Chinese (#1128)
reactjs/zh-hans.react.dev: src/content/learn/add-react-to-an-existing-project.md

---
title: 将 React 添加到现有项目中
---

<Intro>

> 如果想对现有项目添加一些交互，不必将其重写为 React。只需将 React 添加到现有技术栈中，就可以在任何位置渲染交互式的 React 组件。

Author: 我感觉意思应该没有问题，但是表达确实需要调整一下

> 如果想对现有项目添加一些交互，不必将其重写为 React。只需将 React 添加到现有技术栈中，就可以在任何位置渲染交互式的 React 组件。

Author: ```suggestion
如果想对现有项目添加一些交互，不必使用 React 将其重写。只需将 React 添加到现有技术栈中，就可以在任何位置渲染交互式的 React 组件。
```

</Intro>

<Note>

> **你需要安装 [Node.js](https://nodejs.org/zh-cn/) 以进行本地开发。**尽管可以使用 [在线演练场](/learn/installation#try-react) 或简单的 HTML 页面来尝试 React，但实际上大多数用于开发的 JavaScript 工具都需要 Node.js。

Author: ```suggestion
**你需要安装 [Node.js](https://nodejs.org/zh-cn/) 以进行本地开发**。尽管可以使用 [在线演练场](/learn/installation#try-react) 或简单的 HTML 页面来尝试 React，但实际上大多数用于开发的 JavaScript 工具都需要 Node.js。
```

</Note>

> ## 在现有网站的子路由中使用 React {/*using-react-for-an-entire-subroute-of-your-existing-website*/}

Author: ```suggestion
## 在现有网站的子域名中使用 React {/*using-react-for-an-entire-subroute-of-your-existing-website*/}
```

> ## 在现有网站的子路由中使用 React {/*using-react-for-an-entire-subroute-of-your-existing-website*/}

Author: 这个地方感觉还是子路由的意思，但是翻译为子路由好像感觉是怪怪的

> ## 在现有网站的子路由中使用 React {/*using-react-for-an-entire-subroute-of-your-existing-website*/}

Reviewer: 有这么说的，就这样吧

假设你在 `example.com` 部署了一个其他服务端技术（例如 Rails）构建的 Web 应用，但是你又想在 `example.com/some-app/` 部署一个 React 项目。

以下是推荐的设置方式：

> 1. 使用一个 [基于 React 的框架](/learn/start-a-new-react-project) 构建 **应用程序的 React 部分**。

Author: ```suggestion
1. 使用一个 [基于 React 的框架](/learn/start-a-new-react-project) 构建 **应用的 React 部分**。
```

2. **在框架配置中将 `/some-app` 指定为基本路径**（这里有 [Next.js](https://nextjs.org/docs/api-reference/next.config.js/basepath) 与 [Gatsby](https://www.gatsbyjs.com/docs/how-to/previews-deploys-hosting/path-prefix/) 的配置样例）。

> 3. 配置服务器或代理，以便所有位于 `/some-app/` 下的请求都由 React 应用处理。

Author: ```suggestion
3. **配置服务器或代理**，以便所有位于 `/some-app/` 下的请求都由 React 应用处理。
```

> 3. 配置服务器或代理，以便所有位于 `/some-app/` 下的请求都由 React 应用程序处理。

Author: ```suggestion
3. 配置服务器或代理，以便所有位于 `/some-app/` 下的请求都由 React 应用处理。
```

> 这可以确保应用程序的 React 部分可以受益于这些框架中内置的 [最佳实践](/learn/start-a-new-react-project#can-i-use-react-without-a-framework)。

Author: ```suggestion
这可以确保应用的 React 部分可以受益于这些框架中内置的 [最佳实践](/learn/start-a-new-react-project#can-i-use-react-without-a-framework)。
```

> 许多基于 React 的框架都是全栈的，从而可以让你的 React 应用程序充分利用服务器。但是，即使无法或不想在服务器上运行 JavaScript，也可以使用相同的方法。在这种情况下，将 HTML/CSS/JS 导出（Next.js 的 [`next export` output](https://nextjs.org/docs/advanced-features/static-html-export)，Gatsby 的 default）替代为 `/some-app/`。

Author: ```suggestion
许多基于 React 的框架都是全栈的，从而可以让你的 React 应用充分利用服务器。但是，即使无法或不想在服务器上运行 JavaScript，也可以使用相同的方法。在这种情况下，将 HTML/CSS/JS 导出（Next.js 的 [`next export` output](https://nextjs.org/docs/advanced-features/static-html-export)，Gatsby 的 default）替代为 `/some-app/`。
```

## 在现有页面的一部分中使用 React {/*using-react-for-a-part-of-your-existing-page*/}

> 假设有一个其他技术栈（无论是 Rails 这样的服务器技术，还是 Backbone 那样的客户端技术）构建的现有页面，并且想要在该页面的某个位置呈现交互式的 React 组件。这是集成 React 的常见方式——实际上，这也正是多年来大多数 React 用户关注 Meta 的方式！

Author: ```suggestion
假设有一个其他技术栈（无论是 Rails 这样的服务器技术，还是 Backbone 那样的客户端技术）构建的现有页面，并且想要在该页面的某个位置渲染交互式的 React 组件。这是集成 React 的常见方式——实际上，这也正是多年来大多数 React 用户关注 Meta 的方式！
```

> 假设有一个其他技术栈（无论是 Rails 这样的服务器技术，还是 Backbone 那样的客户端技术）构建的现有页面，并且想要在该页面的某个位置渲染交互式的 React 组件。这是集成 React 的常见方式——实际上，这也正是多年来大多数 React 用户关注 Meta 的方式！

Reviewer: ```suggestion
假设有一个其他技术栈（无论是 Rails 这样的服务端技术，还是 Backbone 那样的客户端技术）构建的现有页面，并且想要在该页面的某个位置渲染交互式的 React 组件。这是集成 React 的常见方式——实际上，这也正是多年来大多数情况下 Meta 使用 React 的方式！
```

你可以分两步进行：

> 1. **配置 JavaScript 环境**，以便使用 [JSX 语法](/learn/writing-markup-with-jsx)、 [`import`](https://developer.mozilla.org/zh-CN/docs/Web/JavaScript/Reference/Statements/import) 和 [`export`](https://developer.mozilla.org/zh-CN/docs/Web/JavaScript/Reference/Statements/export) 语法将代码拆分为模块，以及从 [npm](https://www.npmjs.com/) 包注册表中使用包（例如 React）。

Reviewer: ```suggestion
1. **配置 JavaScript 环境**，以便使用 [JSX 语法](/learn/writing-markup-with-jsx)、[`import`](https://developer.mozilla.org/zh-CN/docs/Web/JavaScript/Reference/Statements/import) 和 [`export`](https://developer.mozilla.org/zh-CN/docs/Web/JavaScript/Reference/Statements/export) 语法将代码拆分为模块，以及从 [npm](https://www.npmjs.com/) 包注册表中使用包（例如 React）。
```

> 2. **在想在页面上看到它们的位置呈现 React 组件**。

Author: ```suggestion
2. **在想在页面上看到它们的位置渲染 React 组件**。
```

确切的方法取决于现有的页面设置，因此让我们详细介绍一些细节。

### 步骤 1：设置模块化的 JavaScript 环境 {/*step-1-set-up-a-modular-javascript-environment*/}

模块化的 JavaScript 环境可以让你在单独的文件中编写 React 组件，而不是在一个文件中编写所有的代码。它还可以让你使用其他开发人员在 [npm](https://www.npmjs.com/) 注册表上发布的所有精彩包，包括 React 本身！如何实现这一点取决于你现有的设置：

> * **如果你的应用程序已经分成了使用 `import` 语句的文件，请尝试使用你已有的设置。** 检查在你的 JavaScript 代码中编写 `<div />`是否会导致语法错误。如果导致语法错误，你可能需要使用 Babel 转换你的 JavaScript 代码，并启用 Babel React 预设来使用 JSX。

Reviewer 2: ```suggestion
* **如果你的应用程序已经分成了使用 `import` 语句的文件，请尝试使用你已有的设置。** 检查在你的 JavaScript 代码中编写 `<div />`是否会导致语法错误。如果导致语法错误，你可能需要使用 [Babel](https://babeljs.io/setup) 转换你的 JavaScript 代码，并启用 [Babel React preset](https://babeljs.io/docs/babel-preset-react) 来使用 JSX。
```

这里为什么把链接去掉了？

> * **如果你的应用已经分成了使用 `import` 语句的文件，请尝试使用你已有的设置。** 检查在你的 JavaScript 代码中编写 `<div />`是否会导致语法错误。如果导致语法错误，你可能需要使用 [Babel](https://babeljs.io/setup) 转换你的 JavaScript 代码，并启用 [Babel React preset](https://babeljs.io/docs/babel-preset-react) 来使用 JSX。

Author: ```suggestion
* **如果你的应用已经分成了使用 `import` 语句的文件，请尝试使用你已有的设置**。检查在你的 JavaScript 代码中编写 `<div />`是否会导致语法错误。如果导致语法错误，你可能需要使用 [Babel](https://babeljs.io/setup) 转换你的 JavaScript 代码，并启用 [Babel React preset](https://babeljs.io/docs/babel-preset-react) 来使用 JSX。
```

> * **如果你的应用程序已经分成了使用 `import` 语句的文件，请尝试使用你已有的设置。** 检查在你的 JavaScript 代码中编写 `<div />`是否会导致语法错误。如果导致语法错误，你可能需要使用 [Babel](https://babeljs.io/setup) 转换你的 JavaScript 代码，并启用 [Babel React preset](https://babeljs.io/docs/babel-preset-react) 来使用 JSX。

Author: ```suggestion
* **如果你的应用已经分成了使用 `import` 语句的文件，请尝试使用你已有的设置。** 检查在你的 JavaScript 代码中编写 `<div />`是否会导致语法错误。如果导致语法错误，你可能需要使用 [Babel](https://babeljs.io/setup) 转换你的 JavaScript 代码，并启用 [Babel React preset](https://babeljs.io/docs/babel-preset-react) 来使用 JSX。
```

> * **如果你的应用程序已经分成了使用 `import` 语句的文件，请尝试使用你已有的设置。** 检查在你的 JavaScript 代码中编写 `<div />`是否会导致语法错误。如果导致语法错误，你可能需要使用 [Babel](https://babeljs.io/setup) 转换你的 JavaScript 代码，并启用 [Babel React preset](https://babeljs.io/docs/babel-preset-react) 来使用 JSX。

Author: ```suggestion
* **如果你的应用已经分成了使用 `import` 语句的文件，请尝试使用你已有的设置。** 检查在你的 JavaScript 代码中编写 `<div />`是否会导致语法错误。如果导致语法错误，你可能需要使用 [Babel](https://babeljs.io/setup) 转换你的 JavaScript 代码，并启用 [Babel React preset](https://babeljs.io/docs/babel-preset-react) 来使用 JSX。
```

> * **如果你的应用没有用于编译 JavaScript 模块的配置，请使用 [Vite](https://vitejs.dev/) 进行配置**。Vite 社区维护了与后端框架（包括 Rails、Django 和 Laravel）的 [许多集成项目](https://github.com/vitejs/awesome-vite#integrations-with-backends)。如果你的后端框架没有列出，请 [按照此指南](https://cn.vitejs.dev/guide/backend-integration.html) 手动将 Vite 构建集成到你的后端。

Reviewer: ```suggestion
* **如果你的应用没有用于编译 JavaScript 模块的配置，请使用 [Vite](https://cn.vitejs.dev/) 进行配置**。Vite 社区维护了与后端框架（包括 Rails、Django 和 Laravel）的 [许多集成项目](https://github.com/vitejs/awesome-vite#integrations-with-backends)。如果你的后端框架没有列出，请 [按照此指南](https://cn.vitejs.dev/guide/backend-integration.html) 手动将 Vite 构建集成到你的后端。
```

> * **如果你的应用程序没有用于编译 JavaScript 模块的现有设置，请使用 [Vite](https://vitejs.dev/)进行设置**。Vite 社区维护了与后端框架（包括 Rails、Django 和 Laravel）的 [许多集成](https://github.com/vitejs/awesome-vite#integrations-with-backends)。如果你的后端框架没有列出，请 [按照此指南](https://cn.vitejs.dev/guide/backend-integration.html) 手动将 Vite 构建集成到你的后端。

Author: ```suggestion
* **如果你的应用没有用于编译 JavaScript 模块的现有设置，请使用 [Vite](https://vitejs.dev/)进行设置**。Vite 社区维护了与后端框架（包括 Rails、Django 和 Laravel）的 [许多集成](https://github.com/vitejs/awesome-vite#integrations-with-backends)。如果你的后端框架没有列出，请 [按照此指南](https://cn.vitejs.dev/guide/backend-integration.html) 手动将 Vite 构建集成到你的后端。
```

  <head><title>My app</title></head>
  <body>
    <!-- 你现有的页面内容（在这个例子中，它将被替换） -->
  </body>
</html>
```

```js index.js active
import { createRoot } from 'react-dom/client';

// 清除现有的 HTML 内容
document.body.innerHTML = '<div id="app"></div>';

// 渲染你的 React 组件
const root = createRoot(document.getElementById('app'));
root.render(<h1>Hello, world</h1>);
```

</Sandpack>

如果页面的全部内容都被替换为“Hello, world!”，则一切正常！那么继续阅读。

<Note>

第一次将模块化 JavaScript 环境集成到现有项目中可能会让人感到害怕，但这是值得的！如果遇到困难，请尝试我们的 [社区资源](/community) 或 [Vite Chat](https://chat.vitejs.dev/)。

</Note>

### 步骤 2：在页面的任何位置渲染 React 组件 {/*step-2-render-react-components-anywhere-on-the-page*/}

在上一步中，你将此代码放在主文件的顶部：

```js
import { createRoot } from 'react-dom/client';

// 清除现有的 HTML 内容
document.body.innerHTML = '<div id="app"></div>';

// 渲染你的 React 组件
const root = createRoot(document.getElementById('app'));
root.render(<h1>Hello, world</h1>);
```

当然，你实际上并不想清除现有的 HTML 内容！

那么请删除此代码。

相反，你可能想要在 HTML 中特定的位置渲染 React 组件。打开 HTML 页面（或生成它的服务器模板），并向任何标记添加一个唯一的 [`id`](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/id) 属性，例如：

```html
<!-- ... 你的 HTML 代码某处 ... -->
<nav id="navigation"></nav>
<!-- ... 其他 HTML 代码 ... -->
```

> 这样可以使用 [`document.getElementById`](https://developer.mozilla.org/en-US/docs/Web/API/Document/getElementById) 查找到该 HTML 元素，并将其传递给 [`createRoot`](/reference/react-dom/client/createRoot)，以便可以在其中呈现自己的 React 组件：

Author: ```suggestion
这样可以使用 [`document.getElementById`](https://developer.mozilla.org/en-US/docs/Web/API/Document/getElementById) 查找到该 HTML 元素，并将其传递给 [`createRoot`](/reference/react-dom/client/createRoot)，以便可以在其中渲染自己的 React 组件：
```

</Sandpack>

> 请注意，`index.html` 中的原始 HTML 内容被保留，但你自己的 `NavigationBar` React 组件现在出现在 HTML 的 `<nav id="navigation">` 中。阅读 [`createRoot` 用法文档](/reference/react-dom/client/createRoot#rendering-a-page-partially-built-with-react) 以了解如何在现有 HTML 页面中渲染 React 组件。

Reviewer 3: ```suggestion
请注意 `index.html` 中的原始 HTML 内容是如何保留的，但现在你自己的 `NavigationBar` React 组件出现在 HTML 的 `<nav id="navigation">` 中。阅读 [`createRoot` 用法文档](/reference/react-dom/client/createRoot#rendering-a-page-partially-built-with-react) 以了解如何在现有 HTML 页面中渲染 React 组件。
```

> 当在现有项目中采用 React 时，通常会从小型交互式组件（例如按钮）开始，然后逐渐“向上移动”，直到最终整个页面都由 React 构建。当你实现到某个程度时，我们建议立即迁移到 [一个 React 框架](/learn/start-a-new-react-project)，以充分利用 React 的优势。

Reviewer: ```suggestion
当在现有项目中采用 React 时，通常会从小型交互式组件（例如按钮）开始，然后逐渐“向上移动”，直到最终整个页面都由 React 构建。到那个时候，我们建议立即迁移到 [一个 React 框架](/learn/start-a-new-react-project)，以充分利用 React 的优势。
```

</Sandpack>

> 请注意，`index.html` 中的原始 HTML 内容被保留，但你自己的 `NavigationBar` React 组件现在出现在 HTML 的 `<nav id="navigation">` 中。阅读 [`createRoot` 用法文档](/reference/react-dom/client/createRoot#rendering-a-page-partially-built-with-react) 以了解如何在现有 HTML 页面中呈现 React 组件。

Author: ```suggestion
请注意，`index.html` 中的原始 HTML 内容被保留，但你自己的 `NavigationBar` React 组件现在出现在 HTML 的 `<nav id="navigation">` 中。阅读 [`createRoot` 用法文档](/reference/react-dom/client/createRoot#rendering-a-page-partially-built-with-react) 以了解如何在现有 HTML 页面中渲染 React 组件。
```

当在现有项目中采用 React 时，通常会从小型交互式组件（例如按钮）开始，然后逐渐“向上移动”，直到最终整个页面都由 React 构建。当你实现到某个程度时，我们建议立即迁移到 [一个 React 框架](/learn/start-a-new-react-project)，以充分利用 React 的优势。

> 

Reviewer: ```suggestion
```

> ## 在现有的原生移动应用程序中使用 React Native {/*using-react-native-in-an-existing-native-mobile-app*/}

Author: ```suggestion
## 在现有的原生移动应用中使用 React Native {/*using-react-native-in-an-existing-native-mobile-app*/}
```

> [React Native](https://reactnative.dev/) 也可以逐步集成到现有的原生应用程序中。如果已经有一个现有的 Android（Java 或 Kotlin）或 iOS（Objective-C 或 Swift）原生应用程序，请 [按照本指南](https://reactnative.dev/docs/integration-with-existing-apps) 将 React Native 添加到其中。

Author: ```suggestion
[React Native](https://reactnative.dev/) 也可以逐步集成到现有的原生应用中。如果已经有一个现有的 Android（Java 或 Kotlin）或 iOS（Objective-C 或 Swift）原生应用，请 [按照本指南](https://reactnative.dev/docs/integration-with-existing-apps) 将 React Native 添加到其中。
```
